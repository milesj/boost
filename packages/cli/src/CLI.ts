import { Argv } from '@boost/args';
import { Contract, Predicates } from '@boost/common';
import { CLIOptions } from './types';

export default class CLI extends Contract<CLIOptions> {
  blueprint({ string }: Predicates) {
    return {
      bin: string().required(),
      name: string().required(),
      version: string().required(),
    };
  }

  run(argv: Argv) {}
}
