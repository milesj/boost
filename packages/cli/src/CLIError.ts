import { createScopedError } from '@boost/internal';

const errors = {
  COMMAND_EXISTS: 'A command already exists with the canonical path "%s".',
  COMMAND_INVALID_REGISTER: 'Invalid command type being registered.',
  COMMAND_INVALID_RUN: 'Failed to determine a command to run.',
  COMMAND_INVALID_SUBPATH: 'Sub-command "%s" must start with "%s:".',
  COMMAND_MIXED_DEFAULT:
    'A default command has been registered. Cannot mix default and non-default commands.',
  COMMAND_MIXED_NONDEFAULT:
    'Other commands have been registered. Cannot mix default and non-default commands.',
  COMMAND_NO_PROGRAM: 'No program found. Must be ran within the context of a parent program.',
  COMMAND_NONE_REGISTERED: 'No commands have been registered. At least 1 is required.',
  COMMAND_UNKNOWN: 'Unknown command "%s". Did you mean "%s"?',
  MIDDLEWARE_INVALID: 'Middleware must be a function.',
  NO_NESTED_REACT_RENDER:
    'A React render process already exists. Unable to render current command. This usually occurs by triggering nested program calls.',
  OPTION_RESERVED: 'Option "%s" is a reserved name and cannot be used.',
  PARAMS_RUN_ONLY: 'Parameters must be defined on the `run()` method.',
};

export type CLIErrorCode = keyof typeof errors;

export default createScopedError<CLIErrorCode>('CLI', 'CLIError', errors);
