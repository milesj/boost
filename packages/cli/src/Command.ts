import {
  ArgList,
  PrimitiveType,
  ParserOptions,
  ParamConfigList,
  OptionConfigMap,
} from '@boost/args';
import { Logger } from '@boost/log';
import { msg } from './constants';
import {
  GlobalArgumentOptions,
  Commandable,
  CommandMetadata,
  ExitHandler,
  RunResult,
} from './types';
import getConstructor from './metadata/getConstructor';
import getInheritedOptions from './metadata/getInheritedOptions';
import validateParams from './metadata/validateParams';
import validateOptions from './metadata/validateOptions';
import validateConfig from './metadata/validateConfig';

export default abstract class Command<
  O extends GlobalArgumentOptions = GlobalArgumentOptions,
  P extends PrimitiveType[] = ArgList
> implements Commandable<P> {
  static description: string = '';

  static deprecated: boolean = false;

  static hidden: boolean = false;

  static options: OptionConfigMap = {
    help: {
      description: msg('cli:optionHelpDescription'),
      short: 'h',
      type: 'boolean',
    },
    locale: {
      default: 'en',
      description: msg('cli:optionLocaleDescription'),
      type: 'string',
    },
    version: {
      description: msg('cli:optionVersionDescription'),
      short: 'v',
      type: 'boolean',
    },
  };

  static params: ParamConfigList = [];

  static path: string = '';

  static rest: string[] = [];

  static usage: string | string[] = '';

  help: boolean = false;

  locale: string = 'en';

  version: boolean = false;

  exit!: ExitHandler;

  log!: Logger;

  protected subCommands: CommandMetadata['commands'] = {};

  constructor() {
    const ctor = getConstructor(this);

    validateConfig(this.constructor.name, {
      deprecated: ctor.deprecated,
      description: ctor.description,
      hidden: ctor.hidden,
      path: ctor.path,
      rest: ctor.rest,
      usage: ctor.usage,
    });
    validateOptions(ctor.options);
    validateParams(ctor.params);
  }

  /**
   * Life cycle that is triggered before the command is run.
   */
  bootstrap() {}

  /**
   * Validate and return all metadata registered to this command instance.
   */
  getMetadata(): CommandMetadata {
    const ctor = getConstructor(this);

    return {
      commands: this.subCommands,
      deprecated: ctor.deprecated,
      description: ctor.description,
      hidden: ctor.hidden,
      options: getInheritedOptions(this),
      params: ctor.params,
      path: ctor.path,
      rest: ctor.rest,
      usage: ctor.usage,
    };
  }

  /**
   * Return metadata as options for argument parsing.
   */
  getParserOptions(): ParserOptions<O, P> {
    const { options, params, path } = this.getMetadata();

    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    return {
      commands: [path],
      options,
      params,
    } as ParserOptions<O, P>;
  }

  /**
   * Return the command path (canonical name on the command line).
   */
  getPath(): string {
    return getConstructor(this).path;
  }

  /**
   * Register a sub-command for the current command.
   */
  register(command: Commandable): this {
    const path = this.getPath();
    const subPath = command.getPath();

    if (!subPath.startsWith(path)) {
      throw new Error(`Sub-command "${subPath}" must start with "${path}:".`);
    }

    if (this.subCommands[subPath] && this.subCommands[subPath] !== command) {
      throw new Error(`A command already exists with the canonical path "${subPath}".`);
    }

    this.subCommands[subPath] = command;

    return this;
  }

  /**
   * Executed when the command is being ran.
   */
  abstract async run(...params: P): Promise<RunResult>;
}
