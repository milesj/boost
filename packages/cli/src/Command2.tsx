import React from 'react';
import { Help } from './components/Help2';

export default class Command {
  renderHelp(): React.ReactElement {
    const metadata = this.getMetadata();

    return (
      <Help
        categories={metadata.categories}
        config={metadata}
        header={metadata.path}
        options={metadata.options}
        params={metadata.params}
      />
    );
  }
}
