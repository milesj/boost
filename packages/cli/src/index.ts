/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
