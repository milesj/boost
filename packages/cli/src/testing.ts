/* eslint-disable unicorn/import-index, no-param-reassign, jest/prefer-spy-on */

import React from 'react';
import { render } from 'ink';
import { env } from '@boost/internal';
import { stripAnsi } from '@boost/terminal';
import { mockLogger } from '@boost/log/lib/testing';
import { Command, Program, INTERNAL_OPTIONS, INTERNAL_PARAMS } from './index';
import type {
  ExitCode,
  GlobalOptions,
  PrimitiveType,
  ProgramOptions,
  ProgramStreams,
  TaskContext,
} from './index';

export class MockReadStream {
  isTTY = false;
}

export class MockWriteStream {
  append: boolean = false;

  columns: number = 80;

  output: string;

  constructor(append: boolean = false) {
    this.append = append;
    this.output = '';
  }

  write(string: string) {
    if (this.append) {
      this.output += string;
    } else {
      this.output = string;
    }
  }

  get(): string {
    return this.output;
  }

  on() {}
}

export function mockStreams(): ProgramStreams {
  return ({
    stderr: new MockWriteStream(),
    stdin: new MockReadStream(),
    stdout: new MockWriteStream(),
  } as unknown) as ProgramStreams;
}

export function mockProgram(options?: Partial<ProgramOptions>, streams?: ProgramStreams): Program {
  return new Program(
    {
      bin: 'test',
      name: 'Test',
      version: '0.0.0',
      ...options,
    },
    streams || mockStreams(),
  );
}

export async function renderComponent(
  element: React.ReactElement,
  stripped: boolean = false,
): Promise<string> {
  const stdout = new MockWriteStream();

  await render(element, {
    debug: true,
    experimental: true,
    stdout: (stdout as unknown) as NodeJS.WriteStream,
  });

  const output = stdout.get();

  return stripped ? stripAnsi(output) : output;
}

export async function runCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
  command: Command<O, P>,
  params: P,
  options?: Partial<O>,
): Promise<string> {
  if (options) {
    Object.assign(command, options);

    // @ts-expect-error
    command[INTERNAL_OPTIONS] = {
      help: false,
      locale: 'en',
      version: false,
      ...options,
    };
  }

  command.exit = jest.fn();
  command.log = mockLogger();
  command[INTERNAL_PARAMS] = params;

  const result = await command.run(...params);

  if (!result || typeof result === 'string') {
    return result || '';
  }

  return renderComponent(result);
}

export function runTask<A extends unknown[], R, T extends TaskContext>(
  task: (this: T, ...args: A) => R,
  args: A,
  context?: Partial<T>,
): R {
  const notTestable = (name: string) => () => {
    throw new Error(
      `\`${name}\` is not testable using the \`runTask\` utility. Test using a full program.`,
    );
  };

  const baseContext: TaskContext = {
    exit: jest.fn(),
    help: false,
    locale: 'en',
    log: mockLogger(),
    rest: [] as string[],
    runProgram: notTestable('runProgram'),
    runTask: notTestable('runTask'),
    unknown: {},
    version: false,
  };

  return task.apply(
    {
      ...baseContext,
      ...context,
    } as T,
    args,
  );
}

export async function runProgram(
  program: Program,
  argv: string[],
): Promise<{ code: ExitCode; output: string; outputStripped: string }> {
  if (!(program.streams.stderr instanceof MockWriteStream)) {
    program.streams.stderr = (new MockWriteStream() as unknown) as NodeJS.WriteStream;
  }

  if (!(program.streams.stdout instanceof MockWriteStream)) {
    program.streams.stdout = (new MockWriteStream() as unknown) as NodeJS.WriteStream;
  }

  if (!(program.streams.stdin instanceof MockReadStream)) {
    program.streams.stdin = (new MockReadStream() as unknown) as NodeJS.ReadStream;
  }

  // Ink async rendering never resolves while testing,
  // as it relies on system signals to "exit".
  // So we set this to flag the renderer to avoid awaiting.
  env('CLI_TEST_ONLY', 'true');

  const code = await program.run(argv);

  env('CLI_TEST_ONLY', null);

  const output =
    ((program.streams.stdout as unknown) as MockWriteStream).get() +
    ((program.streams.stderr as unknown) as MockWriteStream).get();

  return {
    code,
    output,
    outputStripped: stripAnsi(output),
  };
}
