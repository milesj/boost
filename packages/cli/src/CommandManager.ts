import { ArgList, PrimitiveType } from '@boost/args';
import { Contract, isObject } from '@boost/common';
import { Event } from '@boost/event';
import { CLIError } from './CLIError';
import {
	Commandable,
	CommandMetadata,
	CommandPath,
	GlobalOptions,
	ProxyCommandConfig,
	ProxyCommandRunner,
} from './types';

export abstract class CommandManager<Options extends object = {}> extends Contract<Options> {
	readonly onAfterRegister = new Event<[CommandPath, Commandable]>('after-register');

	readonly onBeforeRegister = new Event<[CommandPath, Commandable]>('before-register');

	protected commands: CommandMetadata['commands'] = {};

	protected commandAliases: Record<string, string> = {};

	/**
	 * Return a command by registered path, or `null` if not found.
	 */
	getCommand<O extends object = {}, P extends PrimitiveType[] = ArgList>(
		path: CommandPath,
	): Commandable<O, P> | null {
		if (!path) {
			return null;
		}

		const alias = this.commandAliases[path];

		return (alias && this.commands[alias]) || this.commands[path] || null;
	}

	/**
	 * Return a list of all registered command paths (including aliases).
	 */
	getCommandPaths(): CommandPath[] {
		return [...Object.keys(this.commands), ...Object.keys(this.commandAliases)];
	}

	/**
	 * Register a command and its canonical path (must be unique),
	 * otherwise an error is thrown.
	 */
	register(command: Commandable): this;

	register<O extends object, P extends PrimitiveType[]>(
		path: CommandPath,
		config: ProxyCommandConfig<O, P>,
		runner: ProxyCommandRunner<O, P>,
	): this;

	register(commandOrPath: Commandable | CommandPath, config?: unknown, runner?: unknown): this {
		let command: Commandable;

		if (
			typeof commandOrPath === 'string' &&
			typeof config !== 'undefined' &&
			typeof runner === 'function'
		) {
			command = this.createProxyCommand(
				commandOrPath,
				config as ProxyCommandConfig<never, ArgList>,
				runner as ProxyCommandRunner<never, ArgList>,
			);
		} else if (isObject(commandOrPath) && typeof commandOrPath.run === 'function') {
			command = commandOrPath;
		} else {
			throw new CLIError('COMMAND_INVALID_REGISTER');
		}

		const { aliases, path } = command.getMetadata();

		this.onBeforeRegister.emit([path, command]);

		this.checkPath(path);
		this.commands[path] = command;

		aliases.forEach((alias) => {
			this.checkPath(alias);
			this.commandAliases[alias] = path;
		});

		this.onAfterRegister.emit([path, command]);

		return this;
	}

	/**
	 * Check that a command path is valid.
	 */
	protected checkPath(path: CommandPath) {
		if (this.commands[path]) {
			throw new CLIError('COMMAND_DEFINED', [path]);
		}
	}

	protected abstract createProxyCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
		path: CommandPath,
		config: ProxyCommandConfig<O, P>,
		runner: ProxyCommandRunner<O, P>,
	): Commandable<O, P>;
}
