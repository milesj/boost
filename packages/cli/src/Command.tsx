/* eslint-disable max-classes-per-file, @typescript-eslint/consistent-type-assertions */

import React from 'react';
import execa, { Options as ExecaOptions } from 'execa';
import {
	ArgList,
	Arguments,
	Argv,
	MapParamType,
	OptionConfigMap,
	ParamConfigList,
	ParserOptions,
	PrimitiveType,
	UnknownOptionMap,
} from '@boost/args';
import { Blueprint, Predicates } from '@boost/common';
import { LoggerFunction } from '@boost/log';
import { CLIError } from './CLIError';
import { CommandManager } from './CommandManager';
import { INTERNAL_OPTIONS, INTERNAL_PARAMS, INTERNAL_PROGRAM } from './constants';
import { Config } from './decorators/Config';
import { mapCommandMetadata } from './helpers/mapCommandMetadata';
import { getConstructor } from './metadata/getConstructor';
import { getInheritedCategories } from './metadata/getInheritedCategories';
import { getInheritedOptions } from './metadata/getInheritedOptions';
import { globalOptions } from './metadata/globalOptions';
import { validateConfig } from './metadata/validateConfig';
import { validateOptions } from './metadata/validateOptions';
import { validateParams } from './metadata/validateParams';
import { Program } from './Program';
import {
	Categories,
	Commandable,
	CommandMetadata,
	CommandPath,
	ExitCode,
	ExitHandler,
	GlobalOptions,
	ProxyCommandConfig,
	ProxyCommandRunner,
	RunResult,
	TaskContext,
} from './types';

/**
 * Create a proxy command using itself as the super class.
 */
export function createProxyCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
	path: CommandPath,
	{ description, options, params, ...config }: ProxyCommandConfig<O, P>,
	runner: ProxyCommandRunner<O, P>,
): Command<O, P> {
	@Config(path, description, config)
	// eslint-disable-next-line @typescript-eslint/no-use-before-define
	class ProxyCommand extends Command<O, P> {
		run(): Promise<RunResult> | RunResult {
			return runner.call(this, this[INTERNAL_OPTIONS]!, this[INTERNAL_PARAMS]!, this.rest);
		}
	}

	if (options !== undefined) {
		ProxyCommand.options = options;
	}

	if (params !== undefined) {
		ProxyCommand.params = params;
	}

	return new ProxyCommand();
}

export abstract class Command<
		O extends GlobalOptions = GlobalOptions,
		P extends PrimitiveType[] = ArgList,
		Options extends object = {},
	>
	extends CommandManager<Options>
	implements Commandable<O, P>
{
	static aliases: string[] = [];

	static allowUnknownOptions: boolean = false;

	static allowVariadicParams: boolean | string = false;

	static categories: Categories = {};

	static category: string = '';

	static description: string = '';

	static deprecated: boolean = false;

	static hidden: boolean = false;

	static options: OptionConfigMap = globalOptions;

	static params: ParamConfigList = [];

	static path: string = '';

	static usage: string[] | string = '';

	// Args

	/** Value of `--help`, `-h` passed on the command line. */
	help: boolean = false;

	/** Value of `--locale` passed on the command line. */
	locale: string = 'en';

	/** Additional arguments passed after `--` on the command line. */
	rest: string[] = [];

	/** Unknown options passed on the command line. */
	unknown: UnknownOptionMap = {};

	/** Value of `--version`, `-v` passed on the command line. */
	version: boolean = false;

	// Methods

	/** Method to exit the current program, with an optional error code. */
	exit!: ExitHandler;

	/** Method to log to the console using a log level. */
	log!: LoggerFunction;

	// Internals

	/** @internal */
	[INTERNAL_OPTIONS]?: O;

	/** @internal */
	[INTERNAL_PARAMS]?: P;

	/** @internal */
	[INTERNAL_PROGRAM]?: Program;

	constructor(options?: Options) {
		super(options);

		const ctor = getConstructor(this);

		validateConfig(this.constructor.name, {
			aliases: ctor.aliases,
			allowUnknownOptions: ctor.allowUnknownOptions,
			allowVariadicParams: ctor.allowVariadicParams,
			categories: ctor.categories,
			category: ctor.category,
			deprecated: ctor.deprecated,
			description: ctor.description,
			hidden: ctor.hidden,
			path: ctor.path,
			usage: ctor.usage,
		});
		validateOptions(ctor.options);
		validateParams(ctor.params);

		this.onBeforeRegister.listen(this.handleBeforeRegister);
	}

	/**
	 * Validate options passed to the constructor.
	 */
	// Empty blueprint so that sub-classes may type correctly
	blueprint(predicates: Predicates): Blueprint<object> {
		return {};
	}

	/**
	 * Create a React element based on the Help component.
	 */
	async createHelp(): Promise<React.ReactElement> {
		const metadata = this.getMetadata();
		const { Help } = await import('./components/Help');

		return (
			<Help
				categories={metadata.categories}
				commands={mapCommandMetadata(metadata.commands)}
				config={metadata}
				delimiter={this[INTERNAL_PROGRAM]?.options.delimiter}
				header={metadata.path}
				options={metadata.options}
				params={metadata.params}
			/>
		);
	}

	/**
	 * Execute a system native command with the given arguments
	 * and pass the results through a promise. This does *not* execute Boost CLI
	 * commands, use `runProgram()` instead.
	 */
	executeCommand(command: string, args: string[], options: ExecaOptions = {}) /* infer */ {
		const { streams } = this.getProgram();

		return execa(command, args, {
			...streams,
			...options,
		});
	}

	/**
	 * Return the current command class as an arguments object.
	 * Options and params must be parsed first to operate correctly.
	 */
	getArguments(): Arguments<O, P> {
		return {
			command: this.getPath().split(':'),
			errors: [],
			options: (this[INTERNAL_OPTIONS] ?? {}) as O,
			params: (this[INTERNAL_PARAMS] ?? []) as unknown as MapParamType<P>,
			rest: this.rest,
			unknown: this.unknown,
		};
	}

	/**
	 * Validate and return all metadata registered to this command instance.
	 */
	getMetadata(): CommandMetadata {
		const ctor = getConstructor(this);
		const options: OptionConfigMap = {};

		// Since default values for options are represented as class properties,
		// we need to inject the defaults into the argument parsing layer.
		// We can easily do this here and avoid a lot of headache.
		Object.entries(getInheritedOptions(this)).forEach(([option, config]) => {
			options[option] = {
				...config,
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
				default: this[option as keyof this] ?? config.default,
			};
		});

		return {
			aliases: ctor.aliases,
			allowUnknownOptions: ctor.allowUnknownOptions,
			allowVariadicParams: ctor.allowVariadicParams,
			categories: getInheritedCategories(this),
			category: ctor.category,
			commands: this.commands,
			deprecated: ctor.deprecated,
			description: ctor.description,
			hidden: ctor.hidden,
			options,
			params: ctor.params,
			path: ctor.path,
			usage: ctor.usage,
		};
	}

	/**
	 * Return metadata as options for argument parsing.
	 */
	getParserOptions(): ParserOptions<O, P> {
		const { aliases, allowUnknownOptions, allowVariadicParams, options, params, path } =
			this.getMetadata();

		return {
			commands: [path, ...aliases],
			options,
			params,
			unknown: allowUnknownOptions,
			variadic: Boolean(allowVariadicParams),
		} as ParserOptions<O, P>;
	}

	/**
	 * Return the command path (canonical name on the command line).
	 */
	getPath(): string {
		return getConstructor(this).path;
	}

	/**
	 * Render a React element with Ink and output to the configured streams.
	 */
	async render(element: React.ReactElement) {
		return this.getProgram().renderElement(element);
	}

	/**
	 * Run the program within itself, by passing a custom command and argv list.
	 */
	runProgram = (argv: Argv): Promise<ExitCode> => this.getProgram().run(argv, undefined, true);

	/**
	 * Run a task (function) with the defined arguments and
	 * the current command instance bound to the task's context.
	 */
	runTask = <A extends unknown[], R>(
		task: (this: TaskContext<O>, ...args: A) => R,
		...args: A
	): R => {
		// We dont want tasks to have full access to the command
		// and its methods, so recreate a similar but smaller context.
		const context: TaskContext<O> = {
			exit: this.exit,
			log: this.log,
			rest: this.rest,
			unknown: this.unknown,
			...this[INTERNAL_OPTIONS]!,
			runProgram: this.runProgram,
			runTask: this.runTask,
		};

		return task.apply(context, args);
	};

	/**
	 * Executed when the command is being ran.
	 */
	abstract run(...params: P): Promise<RunResult> | RunResult;

	/**
	 * Create a proxy command using itself as the super class.
	 */
	protected createProxyCommand<PO extends GlobalOptions, PP extends PrimitiveType[]>(
		path: CommandPath,
		config: ProxyCommandConfig<PO, PP>,
		runner: ProxyCommandRunner<PO, PP>,
	): Command<PO, PP> {
		return createProxyCommand(path, config, runner);
	}

	/**
	 * Return the program instance or fail.
	 */
	private getProgram(): Program {
		const program = this[INTERNAL_PROGRAM];

		if (!program) {
			throw new CLIError('COMMAND_NO_PROGRAM');
		}

		return program;
	}

	/**
	 * Verify sub-command is prefixed with the correct path.
	 */
	private handleBeforeRegister = (subPath: CommandPath) => {
		const path = this.getPath();

		if (!subPath.startsWith(path)) {
			throw new CLIError('COMMAND_INVALID_SUBPATH', [subPath, path]);
		}
	};
}
