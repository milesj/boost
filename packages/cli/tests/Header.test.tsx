import React from 'react';
import { Header } from '../src';
import { renderToString } from '../src/testing';

jest.mock('term-size');

describe('<Header />', () => {
  it('renders `default`', async () => {
    expect(await renderToString(<Header label="Title" type="default" />)).toMatchSnapshot();
  });

  it('renders `muted`', async () => {
    expect(await renderToString(<Header label="Title" type="muted" />)).toMatchSnapshot();
  });

  it('renders `failure`', async () => {
    expect(await renderToString(<Header label="Title" type="failure" />)).toMatchSnapshot();
  });

  it('renders `success`', async () => {
    expect(await renderToString(<Header label="Title" type="success" />)).toMatchSnapshot();
  });

  it('renders `warning`', async () => {
    expect(await renderToString(<Header label="Title" type="warning" />)).toMatchSnapshot();
  });
});
