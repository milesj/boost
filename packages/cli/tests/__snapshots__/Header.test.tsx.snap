// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders \`default\` 1`] = `
"
[1m[47m[30m TITLE [39m[49m[22m
"
`;

exports[`<Header /> renders \`failure\` 1`] = `
"
[1m[41m[30m TITLE [39m[49m[22m
"
`;

exports[`<Header /> renders \`muted\` 1`] = `
"
[1m[100m[30m TITLE [39m[49m[22m
"
`;

exports[`<Header /> renders \`success\` 1`] = `
"
[1m[42m[30m TITLE [39m[49m[22m
"
`;

exports[`<Header /> renders \`warning\` 1`] = `
"
[1m[43m[30m TITLE [39m[49m[22m
"
`;
